transcode/transcode.go

```diff
@@ -165,6 +156,92 @@ func RunTranscodeProcess(transcodeRequest TranscodeSegmentRequest, streamName st
 		return outputs, segmentsCount, err
 	}
 
+	// transmux received segments from T into a single mp4
+	if transcodeRequest.GenerateMP4 {
+		for rlist, slist := range renditionList.RenditionSegmentTable {
+			// create folder
+			if _, err := os.Stat(TRANSMUX_STORAGE_DIR); os.IsNotExist(err) {
+				os.MkdirAll(TRANSMUX_STORAGE_DIR, 0700)
+			}
+
+			// 1. create a .ts file for a given rendition
+			fileName := filepath.Join(TRANSMUX_STORAGE_DIR, transcodeRequest.RequestID+"_"+rlist)
+			f, err := os.Create(fileName+".ts")
+			if err != nil {
+				log.Log(transcodeRequest.RequestID, "XXX: error creating file", fileName, "err", err)
+				f.Close()
+				continue
+			}
+			defer f.Close()
+
+			// 2. for a given rendition, write all segment indices in ascending order to the single .ts file
+			var totalBytes int64
+			for _, k := range slist.GetSortedSegments() {
+//				log.Log(transcodeRequest.RequestID, "XXX:", k, len(slist.SegmentDataTable[k]))
+				fmt.Println(transcodeRequest.RequestID, "XXX:", k, len(slist.SegmentDataTable[k]))
+				segBytes, err := f.Write(slist.SegmentDataTable[k])
+				if err != nil {
+					log.Log(transcodeRequest.RequestID, "XXX: error writing segment", k, "err", err)
+					f.Close()
+					break
+				}
+				totalBytes = totalBytes + int64(segBytes)
+			}
+			// Verify the total bytes written for a given rendition matches the total # of bytes we received from T
+			renditionIndex := getProfileIndex(transcodeProfiles, rlist)
+			fmt.Println("XXX: ", renditionIndex, "transcodeProfiles", transcodeProfiles, "rlist", rlist, "transcodedStats", transcodedStats)
+			var rendBytesWritten int64 = -1
+			for _, v := range transcodedStats {
+				if v.Name == rlist {
+					rendBytesWritten = v.Bytes
+				}
+				fmt.Println("XXX: ", *v)
+			}
+			if rendBytesWritten != totalBytes {
+				log.Log(transcodeRequest.RequestID, "XXX: bytes written does not match", "file", f.Name(), "bytes expected", transcodedStats[renditionIndex].Bytes, "bytes written", totalBytes) 
+				f.Close()
+				break	
+			}
+
+			// transmux the .ts file into mp4
+			err = ffmpeg.Input(f.Name()).
+				Output(fileName+".mp4", ffmpeg.KwArgs{"c": "copy", "bsf:a": "aac_adtstoasc"}).
+				OverWriteOutput().ErrorToStdOut().Run()
+			if err != nil {
+				log.Log(transcodeRequest.RequestID, "XXX: error transmuxing", "file", f.Name(), "err", err)
+				f.Close()
+				continue
+			}
+
+			f2, err := os.Open(fileName+".mp4")
+			if err != nil {
+				log.Log(transcodeRequest.RequestID, "XXX: error opening mp4", "file", f2.Name(), "err", err)
+				break
+			}
+			defer f2.Close()
+
+			// clean-up intermediary .ts file
+			err = os.Remove(f.Name())
+			if err != nil {
+				log.Log(transcodeRequest.RequestID, "XXX: error removing concat ts file")
+			}
+			
+			// TODO: upload mp4 and delete from local on success
+		        targetMP4URL, err := url.JoinPath(targetTranscodedRenditionOutputURL.String(), rlist)
```

Comment:
you can just do `targetTranscodedRenditionOutputURL.JoinPath(rlist).String()` no error to check then either which is nice